File: research/rofl-trustless-faucet/selffund-poc/selffund-faucet/selffund.py
#!/usr/bin/env python3
"""Self-funding agent for a ROFL marketplace machine — reserve-aware.

Runs inside the ROFL TEE. On a loop it:

  1. Reads our instance record (`roflmarket.Instance`) via rofl-appd — one query
     that yields BOTH the runway (`paid_until`) AND the live per-term rent prices
     (`payment.native.terms = {1:hour, 2:month, 3:year}`). Those terms are the
     exact figures the chain will debit on top-up (rofl-market `pay()` charges
     `terms[term] * count`), so the agent never hardcodes or guesses a price —
     if the provider changes the price, the next cycle sees it; if the offer
     doesn't sell a month, the map simply has no `2` key.
  2. When runway falls below SAFETY_WINDOW_SEC, reads the RentPayer reserve
     (the contract's own native balance, via the local appd `accounts.Balances`
     query — no external RPC, network-agnostic) and DECIDES how long a term to
     buy against the live prices:
         - a whole MONTH if the reserve can afford terms[2],
         - otherwise as many HOURS (terms[1]) as it can afford (1..23 per call).
  3. Submits an `evm.Call` (appd `/rofl/v1/tx/sign-submit`) to
     RentPayer.topUp(term, count), which extends the rental out of the
     *contract's* balance. No human tops it up: the enclave pays itself.

The reserve is funded by anyone sending native tokens to RentPayer's public
`receive()`; the agent only ever spends it on rent for our own immutable
instance (topUp is gated on-chain to this enclave via roflEnsureAuthorizedOrigin).

A hard MIN_TOPUP_INTERVAL_SEC guards against runaway spend if the runway query
goes stale; we never spend the reserve below RESERVE_FLOOR_WEI. If prices or the
reserve can't be read, the agent falls back to a minimal 1-hour top-up (the
chain enforces affordability — an unaffordable top-up simply reverts, no harm).

See research/rofl-trustless-faucet/05-verified-architecture.md.
"""
import hashlib
import http.client
import json
import os
import socket
import sys
import time

import cbor2

APPD_SOCK = os.environ.get("APPD_SOCK", "/run/rofl-appd.sock")
RENTPAYER = os.environ.get("RENTPAYER", "").lower()
RENTPAYER_BARE = RENTPAYER.removeprefix("0x")
PROVIDER_HEX = os.environ.get("PROVIDER_HEX", "005599c1f7807c8baa2eec8ddadc395d9b9b460e21")
INSTANCE_ID_HEX = os.environ.get("INSTANCE_ID_HEX", "")

# Per-term prices are NOT configured here — they are read live from the instance
# record each cycle (payment.native.terms). This is the exact figure the chain
# debits, so the decision tracks price changes automatically and can never
# misprice into a revert.

# Keep at least this much in the reserve untouched (default 0 = spend it all
# to stay alive as long as possible; the reserve is refillable by anyone).
RESERVE_FLOOR_WEI = int(os.environ.get("RESERVE_FLOOR_WEI", "0"))

# Top up when remaining runway drops below this many seconds.
SAFETY_WINDOW_SEC = int(os.environ.get("SAFETY_WINDOW_SEC", "3600"))    # 1h headroom
# Never top up more often than this — runaway guard if the runway query stalls.
MIN_TOPUP_INTERVAL_SEC = int(os.environ.get("MIN_TOPUP_INTERVAL_SEC", "600"))
CHECK_INTERVAL_SEC = int(os.environ.get("CHECK_INTERVAL_SEC", "120"))
# Force one top-up at startup (handy right after deploy when paid_until is short).
FORCE_FIRST_TOPUP = os.environ.get("FORCE_FIRST_TOPUP", "0") == "1"

TERM_HOUR, TERM_MONTH, TERM_YEAR = 1, 2, 3
MAX_TERM_COUNT = 23  # CBOR uint stays one byte; contract requires < 24
TOPUP_SELECTOR = "e59b2e35"  # keccak256("topUp(uint8,uint8)")[:4]


class UnixHTTPConnection(http.client.HTTPConnection):
    def __init__(self, path):
        super().__init__("localhost")
        self._unix_path = path

    def connect(self):
        s = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
        s.connect(self._unix_path)
        self.sock = s


def appd(method, path, body=None):
    conn = UnixHTTPConnection(APPD_SOCK)
    headers = {"Content-Type": "application/json"} if body is not None else {}
    conn.request(method, path, body=json.dumps(body) if body is not None else None,
                 headers=headers)
    resp = conn.getresponse()
    data = resp.read()
    conn.close()
    if resp.status >= 300:
        raise RuntimeError(f"appd {method} {path} -> {resp.status}: {data[:200]!r}")
    return data


def get_app_id():
    return json.loads(appd("GET", "/rofl/v1/app/id"))


def query_instance():
    """Read our instance record (status, paid_until). Canonical CBOR required."""
    args = cbor2.dumps(
        {"provider": bytes.fromhex(PROVIDER_HEX), "id": bytes.fromhex(INSTANCE_ID_HEX)},
        canonical=True,
    )
    out = appd("POST", "/rofl/v1/query", {"method": "roflmarket.Instance", "args": args.hex()})
    return cbor2.loads(bytes.fromhex(json.loads(out)["data"]))


def oasis_addr_from_eth(eth_hex):
    """Derive the 21-byte oasis-runtime address for an Ethereum address, exactly
    as runtime-sdk `Address::from_eth`:
        version(0) || sha512_256(ctx || 0x00 || eth_addr)[:20]
    Validated against the known testnet pair. Network-independent — the same
    derivation holds on testnet and mainnet."""
    eth = bytes.fromhex(eth_hex.lower().removeprefix("0x"))
    ctx = b"oasis-runtime-sdk/address: secp256k1eth"
    h = hashlib.new("sha512_256")
    h.update(ctx + b"\x00" + eth)
    return b"\x00" + h.digest()[:20]


def reserve_wei():
    """RentPayer's own native balance = the rent reserve. Read through the LOCAL
    appd socket via the `accounts.Balances` runtime query — no external RPC and
    no network-specific endpoint, so it works identically on testnet and mainnet.
    The native token is the empty-denomination entry."""
    addr = oasis_addr_from_eth(RENTPAYER_BARE)
    args = cbor2.dumps({"address": addr}, canonical=True)
    out = appd("POST", "/rofl/v1/query", {"method": "accounts.Balances", "args": args.hex()})
    res = cbor2.loads(bytes.fromhex(json.loads(out)["data"]))
    balances = res.get("balances") or {}
    if b"" in balances:               # native denomination
        return int(balances[b""])
    if len(balances) == 1:            # single-denom account → take it
        return int(next(iter(balances.values())))
    return 0


def parse_terms(inst):
    """Extract the live per-term native rent prices from an instance record:
    payment = {"native": {"denomination": b"", "terms": {1: hour, 2: month, ...}}}.
    Returns {term_u8: price_wei} or None if the instance isn't native-paid."""
    pay = inst.get("payment")
    if not isinstance(pay, dict):
        return None
    native = pay.get("native")
    if not isinstance(native, dict):
        return None  # EvmContract-paid instance — out of scope for this agent
    raw = native.get("terms") or {}
    out = {}
    for k, v in raw.items():
        try:
            out[int(k)] = int(v)
        except (TypeError, ValueError):
            continue
    return out


def choose_topup(balance, terms):
    """Pick (term, count) buying the LONGEST affordable term against the live
    on-chain prices, or None if the reserve can't fund even one hour above the
    floor. `terms` is {term_u8: price_wei} read from the instance record."""
    spendable = balance - RESERVE_FLOOR_WEI
    if spendable <= 0:
        return None
    month = terms.get(TERM_MONTH)
    hour = terms.get(TERM_HOUR)
    if month and spendable >= month:
        return (TERM_MONTH, 1)
    if hour and spendable >= hour:
        return (TERM_HOUR, int(min(MAX_TERM_COUNT, spendable // hour)))
    return None


def top_up(term, count):
    calldata = TOPUP_SELECTOR + format(term, "064x") + format(count, "064x")
    out = appd("POST", "/rofl/v1/tx/sign-submit", {
        "encrypt": True,
        "tx": {"kind": "eth", "data": {
            "gas_limit": 250000, "to": RENTPAYER_BARE, "value": "0", "data": calldata,
        }},
    })
    return json.loads(out)


def log(*a):
    print(time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()), *a, flush=True)


def term_name(t):
    return {TERM_HOUR: "hour", TERM_MONTH: "month", TERM_YEAR: "year"}.get(t, str(t))


def main():
    log("selffunder starting. rentpayer=%s instance=%s safety_window=%ds "
        "(prices read live from instance.payment.terms)" % (
            RENTPAYER or "(unset)", INSTANCE_ID_HEX or "(unset)", SAFETY_WINDOW_SEC))
    try:
        log("app id:", get_app_id())
    except Exception as e:
        log("WARN get_app_id:", e)

    last_topup = 0.0
    forced_done = False
    while True:
        try:
            if not RENTPAYER or not INSTANCE_ID_HEX:
                log("WAITING: RENTPAYER and INSTANCE_ID_HEX must be set (post-deploy). Idling.")
                time.sleep(CHECK_INTERVAL_SEC)
                continue

            now = time.time()

            # 1. One instance query yields BOTH runway and the live per-term
            #    prices. If readable, runway drives the trigger; if not, we fall
            #    back to the min-interval timer so a stalled query can neither
            #    block nor run away.
            runway = None
            terms = None
            try:
                inst = query_instance()
                paid_until = inst.get("paid_until")
                if paid_until:
                    runway = int(paid_until) - int(now)
                terms = parse_terms(inst)
                log("instance status=%s paid_until=%s runway=%ss terms=%s" % (
                    inst.get("status"), paid_until, runway, terms))
            except Exception as e:
                log("instance query (non-fatal):", repr(e))

            force = FORCE_FIRST_TOPUP and not forced_done
            if runway is not None:
                due = force or runway < SAFETY_WINDOW_SEC
            else:
                due = force or (now - last_topup >= MIN_TOPUP_INTERVAL_SEC)
            # Hard runaway guard regardless of how `due` was decided.
            if due and not force and (now - last_topup) < MIN_TOPUP_INTERVAL_SEC:
                due = False
                log("within min top-up interval (%ds) — holding" % MIN_TOPUP_INTERVAL_SEC)

            if not due:
                log("runway healthy; no top-up this cycle")
                time.sleep(CHECK_INTERVAL_SEC)
                continue

            # 2. Decide term from the live prices + reserve balance. If either
            #    is unreadable, fall back to a minimal 1-hour top-up: the chain
            #    enforces affordability, so an unaffordable attempt just reverts.
            bal = None
            try:
                bal = reserve_wei()
            except Exception as e:
                log("reserve read failed:", repr(e))

            if bal is not None and terms:
                choice = choose_topup(bal, terms)
                if choice is None:
                    log("RESERVE LOW: %s wei available (floor %s) — can't fund an hour "
                        "at %s. Send native tokens to RentPayer %s to keep it alive."
                        % (bal, RESERVE_FLOOR_WEI, terms, RENTPAYER))
                    time.sleep(CHECK_INTERVAL_SEC)
                    continue
                term, count = choice
                log("TOPPING UP %dx %s (reserve=%s wei, terms=%s) ..." % (
                    count, term_name(term), bal, terms))
            else:
                term, count = TERM_HOUR, 1
                log("prices/reserve unread (bal=%s terms=%s) — fallback 1x hour; "
                    "chain enforces affordability" % (bal, terms))

            res = top_up(term, count)
            forced_done = True
            last_topup = now
            log("top-up submitted, appd result:", res)
        except Exception as e:
            log("ERROR loop:", repr(e))
        time.sleep(CHECK_INTERVAL_SEC)


if __name__ == "__main__":
    sys.exit(main())

File: crates/mining-pool/src/self_fund.rs
//! Autonomous ROFL rent self-top-up agent (Rust port of the verified PoC
//! research/rofl-trustless-faucet/selffund-poc/selffund-faucet/selffund.py).
//!
//! Runs as a tokio task inside the ROFL TEE. The other half of the loop — minting
//! fee-MPT against the wallet surplus and swapping it MPT→ROSE into the RentPayer
//! reservoir — is the existing redemption-watcher `fee_swap` task. This agent
//! spends that reservoir on rent:
//!
//!   boot: retarget RentPayer to OUR current machine (setInstance) — survives a
//!         redeploy to a new instance without redeploying the contract.
//!   loop: read our instance record (runway `paid_until` + live per-term prices)
//!         and, before expiry, read the reservoir balance and submit an app-origin
//!         evm.Call to RentPayer.topUp(term,count) — rent paid from the contract.
//!
//! The app-origin (appd `sign-submit`) is what lets RentPayer's
//! `roflEnsureAuthorizedOrigin` pass; a plain RPC tx from the KMS signer would be
//! rejected. Guards: never top up more often than `min_topup_interval` (runaway
//! guard if the runway query stalls); the chain enforces affordability so an
//! over-ambitious top-up simply reverts. Live appd `/query` + `/tx/sign-submit`
//! paths are validated on testnet (no marketplace instance exists on localnet).

use pool_core::appd;
use pool_core::config::SelfFundConfig;
use alloy::primitives::FixedBytes;
use alloy::sol;
use alloy::sol_types::SolCall;
use anyhow::{bail, Context, Result};
use std::time::{Duration, SystemTime, UNIX_EPOCH};
use tracing::{info, warn};

sol! {
    function topUp(uint8 term, uint8 termCount);
    function setInstance(bytes21 provider, bytes8 instanceId);
}

/// Parsed, validated targeting for the agent.
struct Target {
    socket: String,
    rent_payer: [u8; 20],
    provider: [u8; 21],
    instance_id: [u8; 8],
}

fn parse_hex_n<const N: usize>(s: &str, what: &str) -> Result<[u8; N]> {
    let bytes = hex::decode(s.trim().strip_prefix("0x").unwrap_or(s.trim()))
        .with_context(|| format!("{what}: not valid hex"))?;
    if bytes.len() != N {
        bail!("{what}: expected {N} bytes, got {}", bytes.len());
    }
    let mut out = [0u8; N];
    out.copy_from_slice(&bytes);
    Ok(out)
}

fn now_secs() -> u64 {
    SystemTime::now().duration_since(UNIX_EPOCH).map(|d| d.as_secs()).unwrap_or(0)
}

/// Spawn point. Returns immediately if the agent is misconfigured (logs why and
/// does nothing) — never blocks pool startup.
pub async fn run(socket: String, cfg: SelfFundConfig) {
    // ALL targeting is config + AUTHORITATIVE. instance_id is set from
    // `oasis rofl machine show` at deploy. We deliberately do NOT derive the
    // instance from appd's app-id match: deploying a same-app-id instance is
    // permissionless (the revival/redundancy property), so app-id match can't
    // tell OUR machine from a decoy — trusting it would let an attacker
    // misdirect our top-ups (drain the reservoir / expire our machine). The
    // instance id only changes on redeploy, when the deployer sets it anyway.
    let (rent_payer, provider, instance_id) = match (
        parse_hex_n::<20>(&cfg.rent_payer_address, "rent_payer_address"),
        parse_hex_n::<21>(&cfg.provider_hex, "provider_hex"),
        parse_hex_n::<8>(&cfg.instance_id_hex, "instance_id_hex"),
    ) {
        (Ok(rp), Ok(pv), Ok(iid)) => (rp, pv, iid),
        (rp, pv, iid) => {
            warn!(
                rent_payer = ?rp.err(),
                provider = ?pv.err(),
                instance = ?iid.err(),
                "self-fund: targeting unset — agent idle (set rent_payer_address / provider_hex / instance_id_hex)"
            );
            return;
        }
    };

    // Non-authoritative SAFETY cross-check: list the accepted instances of our
    // app on the provider and sanity-check the configured one. We never act on
    // this — it only surfaces a stale config (our instance missing) or a
    // possible decoy (a same-app-id instance we don't recognise).
    if let Ok(app) = appd::app_id_bytes(&socket).await {
        match appd::discover_instances(&socket, &provider, &app).await {
            Ok(found) => {
                if !found.iter().any(|i| *i == instance_id) {
                    warn!(
                        configured = %hex::encode(instance_id),
                        live = ?found.iter().map(hex::encode).collect::<Vec<_>>(),
                        "self-fund: configured instance is NOT among our app's live accepted instances — stale config after a redeploy? (still using the configured id)"
                    );
                }
                if found.iter().filter(|i| **i != instance_id).count() > 0 {
                    warn!(
                        others = found.len(),
                        "self-fund: other instances share our app id (expected on multi-deploy, but watch for a decoy aimed at our reservoir)"
                    );
                }
            }
            Err(e) => warn!(error = %format!("{e:#}"), "self-fund: instance cross-check query failed (non-fatal)"),
        }
    }

    let target = Target { socket, rent_payer, provider, instance_id };

    let reserve_floor: u128 = cfg.reserve_floor_wei.trim().parse().unwrap_or(0);
    info!(
        rent_payer = %format!("0x{}", hex::encode(target.rent_payer)),
        safety_window_secs = cfg.safety_window_secs,
        "self-fund agent started (prices read live from the instance record)"
    );

    match appd::app_id(&target.socket).await {
        Ok(id) => info!(app_id = %id, "self-fund: appd reachable"),
        Err(e) => warn!(error = %format!("{e:#}"), "self-fund: appd /app/id failed"),
    }

    // Boot: retarget the (persistent) RentPayer to this machine's instance.
    if let Err(e) = target.set_instance().await {
        warn!(error = %format!("{e:#}"), "self-fund: setInstance failed (retried next restart)");
    }

    let mut last_topup = 0u64;
    let mut forced_done = false;
    let mut sleep_secs = cfg.min_check_interval_secs.max(1);
    loop {
        tokio::time::sleep(Duration::from_secs(sleep_secs)).await;
        match target.tick(&cfg, reserve_floor, &mut last_topup, &mut forced_done).await {
            // Adaptive cadence: poll ≈ runway/4, clamped — frequent when scraping
            // for the next hour, rare when a month is paid. Unknown runway or an
            // error → poll soon (min) so we don't go blind.
            Ok(runway) => sleep_secs = adaptive_sleep(runway, &cfg),
            Err(e) => {
                warn!(error = %format!("{e:#}"), "self-fund: tick failed");
                sleep_secs = cfg.min_check_interval_secs.max(1);
            }
        }
    }
}

fn adaptive_sleep(runway: Option<u64>, cfg: &SelfFundConfig) -> u64 {
    let min = cfg.min_check_interval_secs.max(1);
    let max = cfg.max_check_interval_secs.max(min);
    match runway {
        Some(r) => (r / 4).clamp(min, max),
        None => min,
    }
}

impl Target {
    async fn set_instance(&self) -> Result<()> {
        let data = setInstanceCall {
            provider: FixedBytes::<21>::from_slice(&self.provider),
            instanceId: FixedBytes::<8>::from_slice(&self.instance_id),
        }
        .abi_encode();
        let res = appd::sign_submit_eth(&self.socket, self.rent_payer, &data, 200_000).await?;
        info!(result = %res, "self-fund: setInstance submitted (RentPayer retargeted to this machine)");
        Ok(())
    }

    /// Returns the current runway (paid_until - now) in seconds, if known, so the
    /// caller can set the adaptive poll cadence.
    async fn tick(
        &self,
        cfg: &SelfFundConfig,
        reserve_floor: u128,
        last_topup: &mut u64,
        forced_done: &mut bool,
    ) -> Result<Option<u64>> {
        let now = now_secs();

        // 1. One instance query → runway + live per-term prices.
        let args = appd::encode_instance_args(&self.provider, &self.instance_id);
        let inst = match appd::query(&self.socket, "roflmarket.Instance", &args).await {
            Ok(raw) => Some(appd::parse_instance(&raw)?),
            Err(e) => {
                warn!(error = %format!("{e:#}"), "self-fund: instance query failed (non-fatal)");
                None
            }
        };
        let runway = inst.as_ref().and_then(|i| i.paid_until).map(|p| p.saturating_sub(now));
        if let Some(r) = runway {
            info!(runway_secs = r, "self-fund: runway");
        }

        // 2. Decide whether a top-up is due.
        let force = cfg.force_first_topup && !*forced_done;
        let mut due = match runway {
            Some(r) => force || r < cfg.safety_window_secs,
            None => force || now.saturating_sub(*last_topup) >= cfg.min_topup_interval_secs,
        };
        // Hard runaway guard: never within min interval (unless the very first forced run).
        if due && !force && now.saturating_sub(*last_topup) < cfg.min_topup_interval_secs {
            due = false;
            info!("self-fund: within min top-up interval — holding");
        }
        if !due {
            return Ok(runway);
        }

        // 3. Read the reserve + choose the longest affordable term. If reads
        //    fail, fall back to a minimal 1-hour top-up — the chain enforces
        //    affordability, so an unaffordable attempt just reverts.
        let balance = {
            let oaddr = appd::oasis_addr_from_eth(&self.rent_payer);
            let bargs = appd::encode_balances_args(&oaddr);
            match appd::query(&self.socket, "accounts.Balances", &bargs).await {
                Ok(raw) => appd::parse_native_balance(&raw).ok(),
                Err(e) => {
                    warn!(error = %format!("{e:#}"), "self-fund: reserve read failed");
                    None
                }
            }
        };

        // Pick the best-value affordable term (cheap month when flush, 1 hour when
        // scraping; capped at max_topup_term). Prices/reserve unread → fall back
        // to the shortest term; the chain enforces affordability so a too-big
        // attempt just reverts.
        let (term, count) = match (balance, inst.as_ref().map(|i| &i.terms)) {
            (Some(bal), Some(terms)) if !terms.is_empty() => {
                match appd::plan_topup(bal, reserve_floor, terms, cfg.max_topup_term) {
                    Some(tc) => tc,
                    None => {
                        warn!(
                            reserve_wei = bal,
                            rent_payer = %format!("0x{}", hex::encode(self.rent_payer)),
                            "self-fund: RESERVE LOW — can't afford even an hour; fund RentPayer to keep the pool alive"
                        );
                        return Ok(runway);
                    }
                }
            }
            _ => (1, 1), // shortest term
        };

        let data = topUpCall { term, termCount: count }.abi_encode();
        let res = appd::sign_submit_eth(&self.socket, self.rent_payer, &data, 250_000).await?;
        *last_topup = now;
        *forced_done = true;
        info!(term, count, result = %res, "self-fund: top-up submitted");
        Ok(runway)
    }
}
